pycuda/gpuarray.py

```diff
@@ -1106,6 +1098,133 @@ class Info(Record):
+        shape = [src.shape[axis] for axis in axes]
+        src_strides = [src.strides[axis] for axis in axes]
+        dst_strides = [dst.strides[axis] for axis in axes]
+
+        # copy functions require contiguity in minor axis, so add new axis if needed
+        if len(shape) == 0 or src_strides[0] != src.dtype.itemsize or dst_strides[0] != dst.dtype.itemsize:
+            shape[0:0] = [1]
+            src_strides[0:0] = [0]
+            dst_strides[0:0] = [0]
+            axes[0:0] = [np.newaxis]
+
+        # collapse contiguous dimensions
+        # and check that dst is in same order as src
+        i = 1
+        while i < len(shape):
+            if dst_strides[i] < dst_strides[i-1]:
+                raise ValueError("src and dst must have same order")
+            if (src_strides[i-1] * shape[i-1] == src_strides[i] and
+                dst_strides[i-1] * shape[i-1] == dst_strides[i]):
+                shape[i-1:i+1] = [shape[i-1] * shape[i]]
+                del src_strides[i]
+                del dst_strides[i]
+                del axes[i]
+            else:
+                i += 1
+
+    if len(shape) <= 1:
+        if isinstance(src, GPUArray):
+            if isinstance(dst, GPUArray):
+                if async:
+                    drv.memcpy_dtod_async(dst.gpudata, src.gpudata, src.nbytes, stream=stream)
+                else:
+                    drv.memcpy_dtod(dst.gpudata, src.gpudata, src.nbytes)
+            else:
+                # The arrays might be contiguous in the sense of
+                # having no gaps, but the axes could be transposed
+                # so that the order is neither Fortran or C.
+                # So, we attempt to get a contiguous view of dst.
+                dst_flat = dst.ravel(order='K')
+                assert np.byte_bounds(dst) == np.byte_bounds(dst_flat)
+                if async:
+                    drv.memcpy_dtoh_async(dst_flat, src.gpudata, stream=stream)
+                else:
+                    drv.memcpy_dtoh(dst_flat, src.gpudata)
+        else:
+            src = src.ravel(order='K')
+            if async:
+                drv.memcpy_htod(dst.gpudata, src, stream=stream)
+            else:
+                drv.memcpy_htod(dst.gpudata, src)
+        return
+
+    if len(shape) == 2:
+        copy = drv.Memcpy2D()
+    elif len(shape) == 3:
+        copy = drv.Memcpy3D()
+    else:
+        raise ValueError("more than 2 discontiguous axes not supported %s" % (tuple(sorted(axes)),))
+
+    if isinstance(src, GPUArray):
+        copy.set_src_device(src.gpudata)
+    else:
+        copy.set_src_host(src)
+
+    if isinstance(dst, GPUArray):
+        copy.set_dst_device(dst.gpudata)
+    else:
+        copy.set_dst_host(dst)
+
+    copy.width_in_bytes = src.dtype.itemsize*shape[0]
+
+    copy.src_pitch = src_strides[1]
+    copy.dst_pitch = dst_strides[1]
+    copy.height = shape[1]
+    
+    if len(shape) == 2:
+        if async:
+            copy(stream=stream)
```

Comment:
I've reread the page, and I think you're right. PyCUDA's API conflates "use the zero stream" (which means "overlap with nothing device-side") with "don't be asynchronous". That should probably be fixed at some point.